File: hello-dart/lib/client/util/error_handler.dart
// Centralized error handling for the audio stream client.
// Provides consistent error handling, logging, and recovery strategies.
// Matches the Java ErrorHandler interface.

import 'dart:math' as math;
import '../../src/logger.dart';

/// Error types enumeration.
enum ErrorType {
  connectionError('CONNECTION_ERROR'),
  fileIoError('FILE_IO_ERROR'),
  protocolError('PROTOCOL_ERROR'),
  timeoutError('TIMEOUT_ERROR'),
  validationError('VALIDATION_ERROR');

  final String value;
  const ErrorType(this.value);
}

/// Error information class.
class ErrorInfo {
  final ErrorType type;
  final String message;
  final String context;
  final int timestamp;
  final bool recoverable;

  ErrorInfo(
      this.type, this.message, this.context, this.timestamp, this.recoverable);

  @override
  String toString() {
    return 'ErrorInfo{type=${type.value}, message=\'$message\', context=\'$context\', recoverable=$recoverable}';
  }
}

/// Centralized error handling for the audio stream client.
class ErrorHandler {
  static final Map<ErrorType, int> _errorCounts = {
    ErrorType.connectionError: 0,
    ErrorType.fileIoError: 0,
    ErrorType.protocolError: 0,
    ErrorType.timeoutError: 0,
    ErrorType.validationError: 0,
  };

  static void Function(ErrorInfo)? _onErrorCallback;

  // Error reporting

  /// Report an error.
  /// @param type error type
  /// @param message error message
  /// @param context additional context
  /// @param recoverable whether the error is recoverable
  static void reportError(ErrorType type, String message,
      {String context = '', bool recoverable = false}) {
    int timestamp = DateTime.now().millisecondsSinceEpoch;
    ErrorInfo errorInfo =
        ErrorInfo(type, message, context, timestamp, recoverable);

    // Increment error count
    _incrementErrorCount(type);

    // Log the error
    String logMessage =
        '[${_errorTypeToString(type)}] $message - Context: $context';

    if (recoverable) {
      Logger.warn(logMessage);
    } else {
      Logger.error(logMessage);
    }

    // Call callback if set
    if (_onErrorCallback != null) {
      _onErrorCallback!(errorInfo);
    }
  }

  // Error handling strategies

  /// Handle connection error with retry logic.
  /// @param message error message
  /// @param retryCount current retry count (will be incremented)
  /// @param maxRetries maximum number of retries
  /// @returns true if should retry, false otherwise
  static bool handleConnectionError(
      String message, int retryCount, int maxRetries) {
    reportError(ErrorType.connectionError, message,
        context: 'Retry $retryCount/$maxRetries',
        recoverable: retryCount < maxRetries);

    return retryCount < maxRetries;
  }

  /// Handle file I/O error.
  /// @param message error message
  /// @param filePath file path that caused the error
  /// @returns false (file I/O errors are not recoverable)
  static bool handleFileIOError(String message, String filePath) {
    reportError(ErrorType.fileIoError, message,
        context: 'File: $filePath', recoverable: false);
    return false;
  }

  /// Handle protocol error.
  /// @param message error message
  /// @param expectedFormat expected format description
  /// @returns false (protocol errors are not recoverable)
  static bool handleProtocolError(String message,
      {String expectedFormat = ''}) {
    String context =
        expectedFormat.isNotEmpty ? 'Expected: $expectedFormat' : '';
    reportError(ErrorType.protocolError, message,
        context: context, recoverable: false);
    return false;
  }

  /// Handle timeout error.
  /// @param message error message
  /// @param timeoutMs timeout value in milliseconds
  /// @returns false (timeout errors are not recoverable by default)
  static bool handleTimeoutError(String message, int timeoutMs) {
    reportError(ErrorType.timeoutError, message,
        context: 'Timeout: ${timeoutMs} ms', recoverable: false);
    return false;
  }

  // Recovery strategies

  /// Determine if an error should be retried.
  /// @param type error type
  /// @param currentAttempt current attempt number
  /// @param maxAttempts maximum number of attempts
  /// @returns true if should retry
  static bool shouldRetry(ErrorType type, int currentAttempt, int maxAttempts) {
    if (currentAttempt >= maxAttempts) {
      return false;
    }

    // Only connection and timeout errors are retryable
    return type == ErrorType.connectionError || type == ErrorType.timeoutError;
  }

  /// Get retry delay in milliseconds using exponential backoff.
  /// @param attempt attempt number (1-based)
  /// @returns delay in milliseconds
  static int getRetryDelayMs(int attempt) {
    // Exponential backoff: 2^(attempt-1) * 1000 ms
    // Attempt 1: 1s, Attempt 2: 2s, Attempt 3: 4s, etc.
    return (math.pow(2, attempt - 1) * 1000).toInt();
  }

  // Callbacks

  /// Set callback for error events.
  /// @param callback callback function
  static void setOnError(void Function(ErrorInfo) callback) {
    _onErrorCallback = callback;
  }

  // Statistics

  /// Get error count for a specific type.
  /// @param type error type
  /// @returns error count
  static int getErrorCount(ErrorType type) {
    return _errorCounts[type] ?? 0;
  }

  /// Get total error count across all types.
  /// @returns total error count
  static int getTotalErrorCount() {
    return _errorCounts.values.fold(0, (sum, count) => sum + count);
  }

  /// Clear all error counts.
  static void clearErrorCounts() {
    for (ErrorType type in _errorCounts.keys) {
      _errorCounts[type] = 0;
    }
    Logger.debug('Cleared all error counts');
  }

  // Private helper methods

  /// Increment error count for a specific type.
  /// @param type error type
  static void _incrementErrorCount(ErrorType type) {
    _errorCounts[type] = (_errorCounts[type] ?? 0) + 1;
  }

  /// Convert error type to string.
  /// @param type error type
  /// @returns string representation
  static String _errorTypeToString(ErrorType type) {
    return type.value;
  }
}

File: hello-nodejs/src/client/util/ErrorHandler.js
/**
 * Centralized error handling for the audio stream client.
 * Provides consistent error handling, logging, and recovery strategies.
 * Matches the Java ErrorHandler interface.
 */

export class ErrorHandler {
  static DEFAULT_PREFIX = "stream";

  constructor() {
    this.errorCounts = {
      CONNECTION_ERROR: 0,
      FILE_IO_ERROR: 0,
      PROTOCOL_ERROR: 0,
      TIMEOUT_ERROR: 0,
      VALIDATION_ERROR: 0,
    };

    this.onErrorCallback = null;
  }

  // Error reporting

  /**
   * Report an error.
   *
   * @param {string} type - error type
   * @param {string} message - error message
   * @param {string} context - additional context
   * @param {boolean} recoverable - whether the error is recoverable
   */
  reportError(type, message, context = "", recoverable = false) {
    const timestamp = Date.now();
    const errorInfo = { type, message, context, timestamp, recoverable };

    // Increment error count
    this.incrementErrorCount(type);

    // Log the error
    const logMessage = `[${this.errorTypeToString(type)}] ${message} - Context: ${context}`;

    if (recoverable) {
      console.warn(logMessage);
    } else {
      console.error(logMessage);
    }

    // Call callback if set
    if (this.onErrorCallback) {
      this.onErrorCallback(errorInfo);
    }
  }

  // Error handling strategies

  /**
   * Handle connection error with retry logic.
   *
   * @param {string} message - error message
   * @param {number} retryCount - current retry count (will be incremented)
   * @param {number} maxRetries - maximum number of retries
   * @returns {boolean} true if should retry, false otherwise
   */
  handleConnectionError(message, retryCount, maxRetries) {
    this.reportError(
      "CONNECTION_ERROR",
      message,
      `Retry ${retryCount}/${maxRetries}`,
      retryCount < maxRetries,
    );

    return retryCount < maxRetries;
  }

  /**
   * Handle file I/O error.
   *
   * @param {string} message - error message
   * @param {string} filePath - file path that caused the error
   * @returns {boolean} false (file I/O errors are not recoverable)
   */
  handleFileIOError(message, filePath) {
    this.reportError("FILE_IO_ERROR", message, `File: ${filePath}`, false);
    return false;
  }

  /**
   * Handle protocol error.
   *
   * @param {string} message - error message
   * @param {string} expectedFormat - expected format description
   * @returns {boolean} false (protocol errors are not recoverable)
   */
  handleProtocolError(message, expectedFormat = "") {
    const context = expectedFormat ? `Expected: ${expectedFormat}` : "";
    this.reportError("PROTOCOL_ERROR", message, context, false);
    return false;
  }

  /**
   * Handle timeout error.
   *
   * @param {string} message - error message
   * @param {number} timeoutMs - timeout value in milliseconds
   * @returns {boolean} false (timeout errors are not recoverable by default)
   */
  handleTimeoutError(message, timeoutMs) {
    this.reportError(
      "TIMEOUT_ERROR",
      message,
      `Timeout: ${timeoutMs} ms`,
      false,
    );
    return false;
  }

  // Recovery strategies

  /**
   * Determine if an error should be retried.
   *
   * @param {string} type - error type
   * @param {number} currentAttempt - current attempt number
   * @param {number} maxAttempts - maximum number of attempts
   * @returns {boolean} true if should retry
   */
  shouldRetry(type, currentAttempt, maxAttempts) {
    if (currentAttempt >= maxAttempts) {
      return false;
    }

    // Only connection and timeout errors are retryable
    return type === "CONNECTION_ERROR" || type === "TIMEOUT_ERROR";
  }

  /**
   * Get retry delay in milliseconds using exponential backoff.
   *
   * @param {number} attempt - attempt number (1-based)
   * @returns {number} delay in milliseconds
   */
  getRetryDelayMs(attempt) {
    // Exponential backoff: 2^(attempt-1) * 1000 ms
    // Attempt 1: 1s, Attempt 2: 2s, Attempt 3: 4s, etc.
    return Math.floor(Math.pow(2, attempt - 1) * 1000);
  }

  // Callbacks

  /**
   * Set callback for error events.
   *
   * @param {Function} callback - callback function
   */
  setOnError(callback) {
    this.onErrorCallback = callback;
  }

  // Statistics

  /**
   * Get error count for a specific type.
   *
   * @param {string} type - error type
   * @returns {number} error count
   */
  getErrorCount(type) {
    return this.errorCounts[type] || 0;
  }

  /**
   * Get total error count across all types.
   *
   * @returns {number} total error count
   */
  getTotalErrorCount() {
    return Object.values(this.errorCounts).reduce(
      (sum, count) => sum + count,
      0,
    );
  }

  /**
   * Clear all error counts.
   */
  clearErrorCounts() {
    for (const type in this.errorCounts) {
      this.errorCounts[type] = 0;
    }
    console.debug("Cleared all error counts");
  }

  // Private helper methods

  /**
   * Increment error count for a specific type.
   *
   * @param {string} type - error type
   */
  incrementErrorCount(type) {
    if (this.errorCounts.hasOwnProperty(type)) {
      this.errorCounts[type]++;
    }
  }

  /**
   * Convert error type to string.
   *
   * @param {string} type - error type
   * @returns {string} string representation
   */
  errorTypeToString(type) {
    return type;
  }
}
